File: src/ui/info_panel.py
"""Image info panel — EXIF data and file metadata."""

from pathlib import Path

import gi

gi.require_version("Gtk", "4.0")
from gi.repository import Gtk

from PIL import Image
from PIL.ExifTags import TAGS


class InfoPanel(Gtk.ScrolledWindow):
    def __init__(self):
        super().__init__()
        self.add_css_class("info-panel")
        self.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.AUTOMATIC)
        self.set_size_request(280, -1)

        self._box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=8)
        self.set_child(self._box)

    def update(self, path: Path | None):
        while child := self._box.get_first_child():
            self._box.remove(child)

        if path is None:
            return

        path = Path(path)

        self._add_section("File")
        self._add_row("Name", path.name)
        self._add_row("Path", str(path.parent))
        try:
            size = path.stat().st_size
            self._add_row("Size", self._format_size(size))
        except OSError:
            pass

        try:
            img = Image.open(path)
            self._add_section("Image")
            self._add_row("Dimensions", f"{img.width} x {img.height}")
            self._add_row("Format", img.format or "Unknown")
            self._add_row("Mode", img.mode)

            exif_data = self._extract_exif(img)
            if exif_data:
                self._add_section("EXIF")
                for key, value in exif_data.items():
                    self._add_row(key, str(value))
            img.close()
        except Exception:
            pass

    def _add_section(self, title: str):
        label = Gtk.Label(label=title, xalign=0)
        label.set_markup(f"<b>{title}</b>")
        label.set_margin_top(8)
        self._box.append(label)
        sep = Gtk.Separator()
        self._box.append(sep)

    def _add_row(self, key: str, value: str):
        row = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=2)
        key_label = Gtk.Label(label=key, xalign=0)
        key_label.add_css_class("info-label")
        val_label = Gtk.Label(label=value, xalign=0, wrap=True)
        val_label.add_css_class("info-value")
        val_label.set_selectable(True)
        row.append(key_label)
        row.append(val_label)
        self._box.append(row)

    def _extract_exif(self, img: Image.Image) -> dict[str, str]:
        result = {}
        try:
            raw = img._getexif()
            if not raw:
                return result
        except AttributeError:
            return result

        interesting = {
            "Make": "Camera",
            "Model": "Model",
            "LensModel": "Lens",
            "ISOSpeedRatings": "ISO",
            "ExposureTime": "Shutter",
            "FNumber": "Aperture",
            "FocalLength": "Focal Length",
            "DateTimeOriginal": "Date Taken",
        }

        decoded = {}
        for tag_id, val in raw.items():
            tag_name = TAGS.get(tag_id, tag_id)
            decoded[tag_name] = val

        for exif_key, display_name in interesting.items():
            if exif_key in decoded:
                val = decoded[exif_key]
                if exif_key == "ExposureTime" and hasattr(val, "numerator"):
                    val = f"{val.numerator}/{val.denominator}s"
                elif exif_key == "FNumber" and hasattr(val, "numerator"):
                    val = f"f/{float(val):.1f}"
                elif exif_key == "FocalLength" and hasattr(val, "numerator"):
                    val = f"{float(val):.0f}mm"
                result[display_name] = str(val)

        return result

    @staticmethod
    def _format_size(size: int) -> str:
        for unit in ("B", "KB", "MB", "GB"):
            if size < 1024:
                return f"{size:.1f} {unit}" if unit != "B" else f"{size} {unit}"
            size /= 1024
        return f"{size:.1f} TB"

File: src/ui/info_panel.rs
use gtk4::prelude::*;
use gtk4;
use std::path::Path;

pub struct InfoPanel {
    pub container: gtk4::ScrolledWindow,
    content_box: gtk4::Box,
}

impl InfoPanel {
    pub fn new() -> Self {
        let container = gtk4::ScrolledWindow::new();
        container.add_css_class("info-panel");
        container.set_policy(gtk4::PolicyType::Never, gtk4::PolicyType::Automatic);
        container.set_size_request(280, -1);

        let content_box = gtk4::Box::new(gtk4::Orientation::Vertical, 8);
        content_box.set_margin_start(12);
        content_box.set_margin_end(12);
        content_box.set_margin_top(8);
        content_box.set_margin_bottom(8);
        container.set_child(Some(&content_box));

        Self {
            container,
            content_box,
        }
    }

    pub fn update(&self, path: Option<&Path>) {
        // Clear existing content
        while let Some(child) = self.content_box.first_child() {
            self.content_box.remove(&child);
        }

        let path = match path {
            Some(p) => p,
            None => return,
        };

        // File info section
        self.add_section("File");
        if let Some(name) = path.file_name() {
            self.add_row("Name", &name.to_string_lossy());
        }
        if let Some(parent) = path.parent() {
            self.add_row("Path", &parent.display().to_string());
        }
        if let Ok(meta) = std::fs::metadata(path) {
            self.add_row("Size", &format_size(meta.len()));
        }

        // Image dimensions — read without full decode
        if let Ok(reader) = image::ImageReader::open(path) {
            if let Ok((w, h)) = reader.into_dimensions() {
                self.add_section("Image");
                self.add_row("Dimensions", &format!("{} \u{00d7} {}", w, h));
            }
        }

        // EXIF data
        if let Ok(file) = std::fs::File::open(path) {
            let mut bufreader = std::io::BufReader::new(file);
            if let Ok(exif) = exif::Reader::new().read_from_container(&mut bufreader) {
                self.add_section("EXIF");
                let fields = [
                    (exif::Tag::Make, "Camera"),
                    (exif::Tag::Model, "Model"),
                    (exif::Tag::LensModel, "Lens"),
                    (exif::Tag::ISOSpeed, "ISO"),
                    (exif::Tag::ExposureTime, "Shutter"),
                    (exif::Tag::FNumber, "Aperture"),
                    (exif::Tag::FocalLength, "Focal Length"),
                    (exif::Tag::DateTimeOriginal, "Date Taken"),
                ];
                for (tag, label) in fields {
                    if let Some(field) = exif.get_field(tag, exif::In::PRIMARY) {
                        self.add_row(label, &field.display_value().to_string());
                    }
                }
            }
        }
    }

    fn add_section(&self, title: &str) {
        let label = gtk4::Label::new(None);
        label.set_markup(&format!("<b>{}</b>", title));
        label.set_xalign(0.0);
        label.set_margin_top(8);
        self.content_box.append(&label);
        self.content_box
            .append(&gtk4::Separator::new(gtk4::Orientation::Horizontal));
    }

    fn add_row(&self, key: &str, value: &str) {
        let row = gtk4::Box::new(gtk4::Orientation::Vertical, 2);
        let key_label = gtk4::Label::new(Some(key));
        key_label.set_xalign(0.0);
        key_label.add_css_class("info-label");
        let val_label = gtk4::Label::new(Some(value));
        val_label.set_xalign(0.0);
        val_label.set_wrap(true);
        val_label.add_css_class("info-value");
        val_label.set_selectable(true);
        row.append(&key_label);
        row.append(&val_label);
        self.content_box.append(&row);
    }
}

fn format_size(size: u64) -> String {
    let mut s = size as f64;
    for unit in &["B", "KB", "MB", "GB"] {
        if s < 1024.0 {
            return if *unit == "B" {
                format!("{} B", size)
            } else {
                format!("{:.1} {}", s, unit)
            };
        }
        s /= 1024.0;
    }
    format!("{:.1} TB", s)
}
